src/Entity/PrintArticleType.php

```diff
@@ -174,17 +185,64 @@ public function buildThumbnail() {
   /**
    * {@inheritdoc}
    */
+  public function save() {
+    $violations = $this->validate();
+    if ($violations->count()) {
+
+      /** @var \Symfony\Component\Validator\ConstraintViolationInterface $violation */
+      foreach ($violations as $violation) {
+        throw new \Exception($violation->getMessage());
+      }
+    }
+    return parent::save();
+  }
+
+  /**
+   * {@inheritdoc}
+   *
+   * @see \Drupal\Core\Entity\ContentEntityBase::preSave()
+   */
   public function preSave(EntityStorageInterface $storage) {
+    // The entity should not be saved, unless it was validated succesfully.
+    if (!$this->validated) {
+      throw new \LogicException('Entity validation was skipped.');
+    }
+    // We reset validated to FALSE, so any further changes would end up in
+    // checking validation again.
+    else {
+      $this->validated = FALSE;
+    }
+
     parent::preSave($storage);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function validate() {
+    $this->validated = TRUE;
 
     $idms = new IDMS($this->idms);
 
-    /** @var \Symfony\Component\Validator\ConstraintViolationListInterface $errors */
-    $errors = $idms->validate();
+    // Validate IDMS.
+    $idmsViolations = $idms->validate();
+
+    // Validate $this.
+    $printArticleTypeViolations = Validation::createValidator()->validate($this, [
+      new IdmsUniqueTags(),
+    ]);
+
+    $tagViolations = Validation::createValidator()->validate($this->getTags(), [
+      new NotBlank(),
```

Comment:
The validation message on the form is not very user friendly: 
![Screen of missing tags validation error](http://share.undpaul.de/bilder/jh/Add_snippet_template__Site-Install_2017-04-28_10-59-02--dd8dx.png).

We could unify the approach to the one in `\Drupal\thunder_print\Entity\TagMapping::validate` and   `\Drupal\thunder_print\Entity\TagMapping::loadValidatorMetadata`.

Or at least pass a custom message to the notBlank validator.